rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,410 @@
+providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### Database object import rules
+
+Teleport will fetch database objects and store them as resources in the backend,
+using a set of global "database object import rules". At certain points in time,
+the database schema will be read and passed through the import rules. Each
+import rule defines a set of database labels that must match in order for the
+import rule to fire. The rules are processed in order (as defined by the
+`priority` field).
+
+Each import rule defines a set mappings. Individual mapping consists of a set of
+attributes to match, like `name`, `object kind` or `schema`, and a list of
+labels to apply. The expected values for an attribute are specified as a list,
+while the matching uses glob-like semantics, with `*` matching any set of
+characters.
+
+#### Examples
+
+The `widget-prod` import rule applies to databases with `env: prod` label. It
+has the priority 10. If there is a database object that matches the clauses in
+`spec.mappings.match`, the import rule will apply the specified labels to the
+database object (`env: prod`, `product: WidgetMaster3000`).
+
+```yaml
+kind: db_object_import_rule
+version: v1
+metadata:
+  name: rule_widget_prod
+spec:
+  priority: 10
+  db_labels:
+    env: prod
+  mappings:
+    - match:
+        databases:
+          - 'Widget*'
+        db_service_names:
+          - all-things-widget
+        names:
+          - '*sales*'
+        object_kinds:
+          - table
+          - view
+          - procedure
+        protocol_names:
+          - postgres
+        schemas:
+          - widget
+          - sales
+          - public
+          - secret
+      add_labels:
+        env: prod
+        product: WidgetMaster3000
+```
+
+There is a Postgres database with a matching labels. Parsing the schema, a
+number of objects is found, including the following one which matches the import
+rule above (pattern matches are noted):
+
+```yaml
+kind: db_object
+version: v1
+metadata:
+  name: widget-sales
+spec:
+  database: WidgetUltimate # matches 'Widget*'
+  db_service_name: all-things-widget
+  name: widget-sales # matches '*sales*'
+  object_kind: table
+  protocol: postgres
+  schema: sales
+```
+
+After processing, the import rule has applied the labels to the object. As there
+are no other import rules matching this object, it is stored in the backend in
+this form.
+
+```yaml
+kind: db_object
+version: v1
+metadata:
+  name: widget-sales
+  labels:
+    env: prod
+    product: WidgetMaster3000
+spec:
+  database: WidgetUltimate
+  db_service_name: all-things-widget
+  name: widget-sales
+  object_kind: table
+  protocol: postgres
+  schema: sales
+```
+
+Any particular attribute can be omitted from the import rule; an import rule
+with empty `match` part will match all objects.
```

Comment:
I don't see a natural way to require at least one non-empty match clause as it is right now, but I think it may change in light of @greedy52 comment https://github.com/gravitational/teleport/pull/33734#discussion_r1432905494.

Reply:
I think empty-means-all is ok for `scope`. For `match`, it should be explicit.

Imagine the scenario where we add support to a new db object type (say `sequence_names`). If empty-means-all, the auto-provisioned user may automatically get permissions to SEQUENCE after the upgrade to the new version. Then the customer has to manually go edit each import rule to add an empty to `sequence_names` to preserve the same permissions as before.